packages/plugin-ext/src/hosted/node/hosted-plugin-deployer-handler.ts

```diff
@@ -31,47 +32,75 @@ export class HostedPluginDeployerHandler implements PluginDeployerHandler {
     /**
      * Managed plugin metadata backend entries.
      */
-    private readonly currentBackendPluginsMetadata: PluginMetadata[] = [];
+    private currentBackendPluginsMetadata: PluginMetadata[] | undefined;
 
     /**
      * Managed plugin metadata frontend entries.
      */
-    private readonly currentFrontendPluginsMetadata: PluginMetadata[] = [];
+    private currentFrontendPluginsMetadata: PluginMetadata[] | undefined;
 
-    getDeployedFrontendMetadata(): PluginMetadata[] {
-        return this.currentFrontendPluginsMetadata;
+    private backendPluginsMetadataDeferred: Deferred<PluginMetadata[]> | undefined;
+
+    private frontendPluginsMetadataDeferred: Deferred<PluginMetadata[]> | undefined;
+
+    getDeployedFrontendMetadata(): Promise<PluginMetadata[]> {
+        if (!this.currentFrontendPluginsMetadata) {
+            this.frontendPluginsMetadataDeferred = new Deferred<PluginMetadata[]>();
+            return this.frontendPluginsMetadataDeferred.promise;
+        }
+        return Promise.resolve(this.currentFrontendPluginsMetadata);
     }
 
-    getDeployedBackendMetadata(): PluginMetadata[] {
-        return this.currentBackendPluginsMetadata;
+    getDeployedBackendMetadata(): Promise<PluginMetadata[]> {
+        if (!this.currentBackendPluginsMetadata) {
+            this.backendPluginsMetadataDeferred = new Deferred<PluginMetadata[]>();
+            return this.backendPluginsMetadataDeferred.promise;
+        }
+        return Promise.resolve(this.currentBackendPluginsMetadata);
     }
 
     async deployFrontendPlugins(frontendPlugins: PluginDeployerEntry[]): Promise<void> {
+        if (!this.currentFrontendPluginsMetadata) {
+            this.currentFrontendPluginsMetadata = [];
```

Comment:
the code below is async, I wonder whether you should assign after all plugin metadata code is resolved, otherwise the frontend can fetch an empty array

Reply:
How about something like:
```ts
export class HostedPluginDeployerHandler implements PluginDeployerHandler {

    private currentFrontendPluginsMetadata: PluginMetadata[] = [];
    private readonly frontendPluginsMetadataDeferred = new Deferred<void>();

    async getDeployedFrontendMetadata(): Promise<PluginMetadata[]> {
        // await first deploy
        await this.frontendPluginsMetadataDeferred.promise;
        // fetch the last deployed state
        return this.currentFrontendPluginsMetadata; 
    }

    async deployFrontendPlugins(frontendPlugins: PluginDeployerEntry[]): Promise<void> {
        const pluginsMetadata: PluginMetadata[] = []; // reset plugins on each deploy
        for (const plugin of frontendPlugins) {
            const metadata = await this.reader.getPluginMetadata(plugin.path());
            if (metadata) {
                if (pluginsMetadata.some(value => value.model.id === metadata.model.id)) {
                    continue;
                }

                pluginsMetadata.push(metadata);
                this.logger.info(`Deploying frontend plugin "${metadata.model.name}@${metadata.model.version}" from "${metadata.model.entryPoint.frontend || plugin.path()}"`);
            }
        }
        // update plugin sync to avoid parallel mutation of `this.currentFrontendPluginsMetadata` by several deployments
        this.currentFrontendPluginsMetadata = pluginsMetadata; 
        // resolve on first deploy
        this.frontendPluginsMetadataDeferred.resolve(undefined);
    }
```

Reply:
Done, thank you!